pallets/parachain-staking/src/lib.rs

```diff
@@ -0,0 +1,1163 @@
+			Self::deposit_event(Event::DelegationDecreased(delegator, candidate, before, after));
+			Ok(().into())
+		}
+
+		/// Unlock rewards from staking which are avaible for the current block
+		/// number. Checks whether the `RewardLocks` BTreeMap for the target has
+		/// entries less or equal than the current block number.
+		///
+		/// NOTE: Unlocking automatically occurs in `note_author` after a block
+		/// author has produced a block. If you assume that a specific collator
+		/// candidate (and their  corresponding delegators) is an active
+		/// collator for every round, it would be unnecessary to ever call
+		/// `unlock_rewards` for such collator and their delegators.
+		#[pallet::weight(0)]
+		pub fn unlock_rewards(
+			origin: OriginFor<T>,
+			// TODO: Better to use Lookup?
+			// target: <T::Lookup as StaticLookup>::Source
+			target: T::AccountId,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			// let target = T::Lookup::lookup(target)?;
+
+			let now: T::BlockNumber = <frame_system::Pallet<T>>::block_number();
+
+			let lock = <RewardLocks<T>>::get(&target);
+			ensure!(!lock.is_empty(), Error::<T>::RewardLockDNE);
+
+			Self::do_update_reward_locks(&target, lock, now);
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn is_delegator(acc: &T::AccountId) -> bool {
+			<DelegatorState<T>>::get(acc).is_some()
+		}
+
+		pub fn is_candidate(acc: &T::AccountId) -> bool {
+			<CollatorState<T>>::get(acc).is_some()
+		}
+
+		pub fn is_selected_candidate(acc: &T::AccountId) -> bool {
+			<SelectedCandidates<T>>::get().binary_search(acc).is_ok()
+		}
+
+		// ensure candidate is active before calling
+		fn update_active(candidate: T::AccountId, total: BalanceOf<T>) {
+			let mut candidates = <CandidatePool<T>>::get();
+			candidates.upsert(Bond {
+				owner: candidate,
+				amount: total,
+			});
+
+			<CandidatePool<T>>::put(candidates);
+		}
+
+		fn compute_block_issuance(
+			collator_stake: BalanceOf<T>,
+			delegator_stake: BalanceOf<T>,
+		) -> (BalanceOf<T>, BalanceOf<T>) {
+			let config = <InflationConfig<T>>::get();
+			config.block_issuance::<T>(collator_stake, delegator_stake)
+		}
+
+		fn delegator_revokes_collator(acc: T::AccountId, collator: T::AccountId) -> DispatchResultWithPostInfo {
+			let mut delegator = <DelegatorState<T>>::get(&acc).ok_or(Error::<T>::DelegatorDNE)?;
+			let old_total = delegator.total;
+			let remaining = delegator
+				.rm_delegation(collator.clone())
+				.ok_or(Error::<T>::DelegationDNE)?;
+			// edge case; if no delegations remaining, leave set of delegators
+			if delegator.delegations.is_empty() {
+				// leave the set of delegators because no delegations left
+				Self::delegator_leaves_collator(acc.clone(), collator)?;
+				<DelegatorState<T>>::remove(&acc);
+				Self::deposit_event(Event::DelegatorLeft(acc, old_total));
+				return Ok(().into());
+			}
+			ensure!(remaining >= T::MinDelegatorStk::get(), Error::<T>::NomBondBelowMin);
```

Comment:
This `ensure!` should never fail, right? Better to highlight that. If it does, what happens to the partially updated storage?

Reply:
It cannot fail if `MinDelegatorStk == MinDelegation`. I will still mention this in a comment. Good catch, had to look it up to be sure.
```rust
/// Minimum stake for any registered on-chain account to delegate
type MinDelegation: Get<BalanceOf<Self>>;
/// Minimum stake for any registered on-chain account to become a delegator
type MinDelegatorStk: Get<BalanceOf<Self>>;
```

Reply:
Fixed in af20eb0